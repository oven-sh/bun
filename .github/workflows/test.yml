name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
