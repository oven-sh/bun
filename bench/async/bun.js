import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
