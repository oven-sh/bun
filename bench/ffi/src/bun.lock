{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
