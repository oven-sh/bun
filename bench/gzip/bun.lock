{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
