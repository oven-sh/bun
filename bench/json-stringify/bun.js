import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
