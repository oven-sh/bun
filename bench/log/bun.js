import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
