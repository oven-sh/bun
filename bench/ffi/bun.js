import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
