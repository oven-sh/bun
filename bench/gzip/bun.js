import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
