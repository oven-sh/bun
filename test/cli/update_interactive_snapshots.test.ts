import { describe, expect, it } from "bun:test";
import { bunEnv, bunExe, tempDirWithFiles } from "harness";

describe("bun update --interactive snapshots", () => {
  it("should not crash with various package name lengths", async () => {
    const dir = tempDirWithFiles("update-interactive-snapshot-test", {
      "package.json": JSON.stringify({
        name: "test-project",
        version: "1.0.0",
        dependencies: {
          "short": "1.0.0",
          "react": "17.0.2",
          "really-long-package-name-for-testing": "1.0.0",
          "@scoped/package": "1.0.0",
          "@organization/extremely-long-scoped-package-name": "1.0.0",
        },
        devDependencies: {
          "dev-pkg": "1.0.0",
          "super-long-dev-package-name-for-testing": "1.0.0",
          "typescript": "4.8.0",
        },
        peerDependencies: {
          "peer-pkg": "1.0.0",
          "very-long-peer-dependency-name": "1.0.0",
        },
        optionalDependencies: {
          "optional-pkg": "1.0.0",
          "long-optional-dependency-name": "1.0.0",
        },
      }),
    });

    // Test that the command doesn't crash with mixed package lengths
    const result = await Bun.spawn({
      cmd: [bunExe(), "update", "--interactive", "--dry-run"],
      cwd: dir,
      env: bunEnv,
      stdin: "pipe",
      stdout: "pipe",
      stderr: "pipe",
    });

    // Send 'n' to exit without selecting anything
    result.stdin.write("n\n");
    result.stdin.end();

    const stdout = await new Response(result.stdout).text();
    const stderr = await new Response(result.stderr).text();

    // Replace version numbers and paths to avoid flakiness
    const normalizedOutput = normalizeOutput(stdout);

    // The output should show proper column spacing and formatting
    expect(normalizedOutput).toMatchSnapshot("update-interactive-no-crash");

    // Should not crash or have formatting errors
    expect(stderr).not.toContain("panic");
    expect(stderr).not.toContain("underflow");
    expect(stderr).not.toContain("overflow");
  });

  it("should handle extremely long package names without crashing", async () => {
    const veryLongName = "a".repeat(80);
    const dir = tempDirWithFiles("update-interactive-long-names", {
      "package.json": JSON.stringify({
        name: "test-project",
        version: "1.0.0",
        dependencies: {
          [veryLongName]: "1.0.0",
          "regular-package": "1.0.0",
        },
      }),
    });

    const result = await Bun.spawn({
      cmd: [bunExe(), "update", "--interactive", "--dry-run"],
      cwd: dir,
      env: bunEnv,
      stdin: "pipe",
      stdout: "pipe",
      stderr: "pipe",
    });

    result.stdin.write("n\n");
    result.stdin.end();

    const stdout = await new Response(result.stdout).text();
    const stderr = await new Response(result.stderr).text();

    const normalizedOutput = normalizeOutput(stdout);

    // Should not crash
    expect(normalizedOutput).toMatchSnapshot("update-interactive-long-names");
    expect(stderr).not.toContain("panic");
    expect(stderr).not.toContain("underflow");
  });

  it("should handle complex version strings without crashing", async () => {
    const dir = tempDirWithFiles("update-interactive-complex-versions", {
      "package.json": JSON.stringify({
        name: "test-project",
        version: "1.0.0",
        dependencies: {
          "package-with-long-version": "1.0.0-alpha.1.2.3.4.5.6.7.8.9.10.11.12",
          "package-with-prerelease": "1.0.0-beta.1+build.12345",
          "package-with-short-version": "1.0.0",
        },
      }),
    });

    const result = await Bun.spawn({
      cmd: [bunExe(), "update", "--interactive", "--dry-run"],
      cwd: dir,
      env: bunEnv,
      stdin: "pipe",
      stdout: "pipe",
      stderr: "pipe",
    });

    result.stdin.write("n\n");
    result.stdin.end();

    const stdout = await new Response(result.stdout).text();
    const stderr = await new Response(result.stderr).text();

    const normalizedOutput = normalizeOutput(stdout);

    // Should not crash
    expect(normalizedOutput).toMatchSnapshot("update-interactive-complex-versions");
    expect(stderr).not.toContain("panic");
    expect(stderr).not.toContain("underflow");
  });
});

function normalizeOutput(output: string): string {
  // Remove Bun version to avoid test flakiness
  let normalized = output.replace(/bun update --interactive v\d+\.\d+\.\d+[^\n]*/g, "bun update --interactive vX.X.X");

  // Normalize any absolute paths
  normalized = normalized.replace(/\/tmp\/[^\/\s]+/g, "/tmp/test-dir");

  // Remove ANSI color codes for cleaner snapshots
  normalized = normalized.replace(/\x1b\[[0-9;]*m/g, "");

  // Remove progress indicators and timing info
  normalized = normalized.replace(/[\r\n]*\s*\([0-9.]+ms\)/g, "");

  // Normalize whitespace
  normalized = normalized.replace(/\r\n/g, "\n");

  return normalized.trim();
}
