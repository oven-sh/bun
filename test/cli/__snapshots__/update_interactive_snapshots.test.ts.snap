// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`bun update --interactive snapshots should format mixed package types with proper spacing: update-interactive-formatting 1`] = `"bun update --interactive vX.X.X"`;

exports[`bun update --interactive snapshots should not crash with various package name lengths: update-interactive-no-crash 1`] = `"bun update --interactive vX.X.X"`;

exports[`bun update --interactive snapshots should handle extremely long package names without crashing: update-interactive-long-names 1`] = `"bun update --interactive vX.X.X"`;

exports[`bun update --interactive snapshots should handle complex version strings without crashing: update-interactive-complex-versions 1`] = `"bun update --interactive vX.X.X"`;
